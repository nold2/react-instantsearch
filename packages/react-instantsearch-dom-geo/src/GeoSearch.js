import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { LatLngPropType } from './propTypes';
import Provider from './Provider';
import GoogleMaps from './GoogleMaps';

class GeoSearch extends Component {
  static propTypes = {
    google: PropTypes.object.isRequired,
    children: PropTypes.func.isRequired,
    initialZoom: PropTypes.number,
    initialPosition: LatLngPropType,
    mapOptions: PropTypes.object,
  };

  static defaultProps = {
    initialZoom: 1,
    initialPosition: {
      lat: 0,
      lng: 0,
    },
    mapOptions: {},
  };

  createBoundingBoxFromHits(hits) {
    const { google } = this.props;

    const latLngBounds = hits.reduce(
      (acc, hit) => acc.extend(hit._geoloc),
      new google.maps.LatLngBounds()
    );

    return {
      northEast: latLngBounds.getNorthEast().toJSON(),
      southWest: latLngBounds.getSouthWest().toJSON(),
    };
  }

  renderProviderChildren = ({
    hits,
    currentRefinement,
    position,
    isRefineOnMapMove,
    hasMapMoveSinceLastRefine,
    refine,
    setMapMoveSinceLastRefine,
  }) => {
    const {
      google,
      initialZoom,
      initialPosition,
      mapOptions,
      children,
    } = this.props;

    const boundingBox =
      !currentRefinement && Boolean(hits.length)
        ? this.createBoundingBoxFromHits(hits)
        : currentRefinement;

    return (
      <GoogleMaps
        testID="GoogleMaps"
        google={google}
        initialZoom={initialZoom}
        initialPosition={initialPosition}
        mapOptions={mapOptions}
        isRefineOnMapMove={isRefineOnMapMove}
        hasMapMoveSinceLastRefine={hasMapMoveSinceLastRefine}
        boundingBox={boundingBox}
        position={position}
        refine={refine}
        setMapMoveSinceLastRefine={setMapMoveSinceLastRefine}
      >
        {children({ hits })}
      </GoogleMaps>
    );
  };

  render() {
    return <Provider testID="Provider">{this.renderProviderChildren}</Provider>;
  }
}

export default GeoSearch;
