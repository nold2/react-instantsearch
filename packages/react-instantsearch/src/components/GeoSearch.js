/* eslint-disable import/no-extraneous-dependencies */
/* global google */

import { isEqual } from 'lodash';
import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
  GoogleMap,
  Marker,
  withGoogleMap,
  withScriptjs,
} from 'react-google-maps';

class GeoSearch extends Component {
  static propTypes = {
    hits: PropTypes.arrayOf(PropTypes.object).isRequired,
    isRefinedWithMap: PropTypes.bool.isRequired,
    refine: PropTypes.func.isRequired,
    position: PropTypes.object,
    currentRefinement: PropTypes.object,
  };

  state = {
    isRefineOnMapMove: true,
    hasMapMoveSinceLastRefine: false,
  };

  isMapAlreadyLoaded = false;
  isPendingRefine = false;
  isUserInteraction = true;

  lastRefinePosition = null;
  lastRefineBoundingBox = null;

  createRef = c => (this.element = c);

  componentDidMount() {
    this.fitViewToBounds();
  }

  componentDidUpdate() {
    const { position, currentRefinement } = this.props;

    // It's not possible to put this in the connector
    // because we can't we have no way to properly:
    // - 1. hook on the update
    // - 2. trigger a proper upadte
    const positionChangedSinceLastRefine =
      Boolean(position) &&
      Boolean(this.lastRefinePosition) &&
      !isEqual(position, this.lastRefinePosition);

    const boundingBoxChangedSinceLastRefine =
      !currentRefinement &&
      Boolean(this.lastRefineBoundingBox) &&
      !isEqual(currentRefinement, this.lastRefineBoundingBox);

    this.lastRefinePosition = position || null;
    this.lastRefineBoundingBox = currentRefinement || null;

    if (positionChangedSinceLastRefine || boundingBoxChangedSinceLastRefine) {
      this.setState(() => ({
        hasMapMoveSinceLastRefine: false,
      }));
    } else {
      this.fitViewToBounds();
    }
  }

  onMapChange = () => {
    const { isRefineOnMapMove } = this.state;

    if (this.isMapAlreadyLoaded && this.isUserInteraction) {
      this.setState(() => ({
        hasMapMoveSinceLastRefine: true,
      }));

      if (isRefineOnMapMove) {
        this.isPendingRefine = true;
      }
    }
  };

  onMapIdle = () => {
    this.isMapAlreadyLoaded = true;

    if (this.isUserInteraction && this.isPendingRefine) {
      this.refineWithMap();

      this.isPendingRefine = false;
    }
  };

  onToggle = () =>
    this.setState(prevState => ({
      isRefineOnMapMove: !prevState.isRefineOnMapMove,
    }));

  refineWithMap = () => {
    const { refine } = this.props;

    const ne = this.element.getBounds().getNorthEast();
    const sw = this.element.getBounds().getSouthWest();

    refine({
      northEast: { lat: ne.lat(), lng: ne.lng() },
      southWest: { lat: sw.lat(), lng: sw.lng() },
    });

    this.setState(() => ({
      hasMapMoveSinceLastRefine: false,
    }));
  };

  clearMapRefinement = () => {
    const { refine } = this.props;

    refine();

    this.setState(() => ({
      hasMapMoveSinceLastRefine: false,
    }));
  };

  fitViewToBounds() {
    const { hits, isRefinedWithMap } = this.props;
    const { hasMapMoveSinceLastRefine } = this.state;
    const isFitBoundsEnable =
      Boolean(hits.length) && !hasMapMoveSinceLastRefine && !isRefinedWithMap;

    if (isFitBoundsEnable) {
      this.isUserInteraction = false;
      this.element.fitBounds(
        hits.reduce(
          (acc, item) =>
            acc.extend({ lat: item._geoloc.lat, lng: item._geoloc.lng }),
          new google.maps.LatLngBounds()
        )
      );
      this.isUserInteraction = true;
    }
  }

  render() {
    const { hits, isRefinedWithMap } = this.props;
    const { isRefineOnMapMove, hasMapMoveSinceLastRefine } = this.state;

    return (
      <div>
        <GoogleMap
          ref={this.createRef}
          defaultZoom={8}
          defaultCenter={{ lat: -34.397, lng: 150.644 }}
          options={{
            mapTypeControl: false,
            fullscreenControl: false,
            streetViewControl: false,
            clickableIcons: false,
            zoomControlOptions: {
              position: google.maps.ControlPosition.LEFT_TOP,
            },
          }}
          onDragStart={this.onMapChange}
          onCenterChanged={this.onMapChange}
          onZoomChanged={this.onMapChange}
          onIdle={this.onMapIdle}
        >
          {hits.map(item => (
            <Marker key={item.objectID} position={item._geoloc} />
          ))}
        </GoogleMap>

        {isRefineOnMapMove || !hasMapMoveSinceLastRefine ? (
          <label>
            <input
              type="checkbox"
              checked={isRefineOnMapMove}
              onChange={this.onToggle}
            />
            Search as I move the map
          </label>
        ) : (
          <button onClick={this.refineWithMap}>Redo search here</button>
        )}

        {isRefinedWithMap && (
          <button onClick={this.clearMapRefinement}>
            Clear map refinement
          </button>
        )}
      </div>
    );
  }
}

export default withScriptjs(withGoogleMap(GeoSearch));
