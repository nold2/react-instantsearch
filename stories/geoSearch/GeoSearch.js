import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Connector from './Connector';
import GoogleMap from './GoogleMap';

class GeoSearch extends Component {
  static propTypes = {
    google: PropTypes.object.isRequired,
    children: PropTypes.func.isRequired,
    initialZoom: PropTypes.number,
    initialPosition: PropTypes.shape({
      lat: PropTypes.number.isRequired,
      lng: PropTypes.number.isRequired,
    }),
  };

  static defaultProps = {
    initialZoom: 1,
    initialPosition: {
      lat: 0,
      lng: 0,
    },
  };

  renderInnerChildren = ({ hits, currentRefinement, position, refine }) => {
    const { google, children, initialZoom, initialPosition } = this.props;

    const hitsLatLngBounds = hits.reduce(
      (acc, item) =>
        acc.extend({ lat: item._geoloc.lat, lng: item._geoloc.lng }),
      new google.maps.LatLngBounds()
    );

    const hitsBoundingBox = {
      northEast: {
        lat: hitsLatLngBounds.getNorthEast().lat(),
        lng: hitsLatLngBounds.getNorthEast().lng(),
      },
      southWest: {
        lat: hitsLatLngBounds.getSouthWest().lat(),
        lng: hitsLatLngBounds.getSouthWest().lng(),
      },
    };

    return (
      <GoogleMap
        google={google}
        boundingBox={currentRefinement || hitsBoundingBox}
        initialZoom={initialZoom}
        initialPosition={position || initialPosition}
        refine={refine}
      >
        {children({ hits })}
      </GoogleMap>
    );
  };

  render() {
    return (
      // prettier-ignore
      <Connector>
        {this.renderInnerChildren}
      </Connector>
    );
  }
}

export default GeoSearch;
