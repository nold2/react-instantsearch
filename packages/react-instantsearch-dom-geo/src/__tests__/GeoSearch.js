import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import { createFakeGoogleReference } from '../../test/fakeGoogleMaps';
import GeoSearch from '../GeoSearch';

Enzyme.configure({ adapter: new Adapter() });

describe('GeoSearch', () => {
  const ShallowWapper = ({ children }) => children;

  const defaultProps = {
    google: createFakeGoogleReference(),
  };

  const defaultRenderProvidedProps = {
    hits: [],
  };

  const renderProps = ({ props, renderProvidedProps, children = () => null }) =>
    shallow(<GeoSearch {...props}>{children}</GeoSearch>)
      .find('[testID="Provider"]')
      .props()
      .children(renderProvidedProps);

  it('expect to render', () => {
    const children = jest.fn(() => <div>Hello this is the children</div>);

    const props = {
      ...defaultProps,
    };

    const renderProvidedProps = {
      ...defaultRenderProvidedProps,
    };

    const renderPropsWrapper = shallow(
      <ShallowWapper>
        {renderProps({ props, renderProvidedProps, children })}
      </ShallowWapper>
    );

    expect(renderPropsWrapper).toMatchSnapshot();
    expect(children).toHaveBeenCalledTimes(1);
    expect(children).toHaveBeenCalledWith({
      hits: [],
    });
  });

  it('expect to render with initialZoom & initialPosition', () => {
    const props = {
      ...defaultProps,
      initialZoom: 8,
      initialPosition: {
        lat: 10,
        lng: 12,
      },
    };

    const renderProvidedProps = {
      ...defaultRenderProvidedProps,
    };

    const renderPropsWrapper = shallow(
      <ShallowWapper>
        {renderProps({ props, renderProvidedProps })}
      </ShallowWapper>
    );

    const googleMapProps = renderPropsWrapper
      .find('[testID="GoogleMaps"]')
      .props();

    expect(googleMapProps.initialZoom).toBe(8);
    expect(googleMapProps.initialPosition).toEqual({
      lat: 10,
      lng: 12,
    });
  });

  it('expect to render with position', () => {
    const props = {
      ...defaultProps,
    };

    const renderProvidedProps = {
      ...defaultRenderProvidedProps,
      position: {
        lat: 10,
        lng: 12,
      },
    };

    const renderPropsWrapper = shallow(
      <ShallowWapper>
        {renderProps({ props, renderProvidedProps })}
      </ShallowWapper>
    );

    const googleMapProps = renderPropsWrapper
      .find('[testID="GoogleMaps"]')
      .props();

    expect(googleMapProps.position).toEqual({
      lat: 10,
      lng: 12,
    });
  });

  describe('boundingBox', () => {
    it('expect to use hits when currentRefinement is not defined and hits are not empty', () => {
      const google = createFakeGoogleReference();

      google.maps.LatLngBounds.mockImplementation(() => ({
        extend: jest.fn().mockReturnThis(),
        getNorthEast: () => ({
          toJSON: () => ({
            lat: 10,
            lng: 10,
          }),
        }),
        getSouthWest: () => ({
          toJSON: () => ({
            lat: 14,
            lng: 14,
          }),
        }),
      }));

      const props = {
        ...defaultProps,
        google,
      };

      const renderProvidedProps = {
        ...defaultRenderProvidedProps,
        hits: [
          { _geoloc: { lat: 10, lng: 12 } },
          { _geoloc: { lat: 12, lng: 14 } },
        ],
      };

      const renderPropsWrapper = shallow(
        <ShallowWapper>
          {renderProps({ props, renderProvidedProps })}
        </ShallowWapper>
      );

      const googleMapProps = renderPropsWrapper
        .find('[testID="GoogleMaps"]')
        .props();

      expect(googleMapProps.boundingBox).toEqual({
        northEast: {
          lat: 10,
          lng: 10,
        },
        southWest: {
          lat: 14,
          lng: 14,
        },
      });
    });

    it("expect to use currentRefinement when it's defined and hits are empty", () => {
      const props = {
        ...defaultProps,
      };

      const renderProvidedProps = {
        ...defaultRenderProvidedProps,
        currentRefinement: {
          northEast: {
            lat: 10,
            lng: 12,
          },
          southWest: {
            lat: 12,
            lng: 14,
          },
        },
      };

      const renderPropsWrapper = shallow(
        <ShallowWapper>
          {renderProps({ props, renderProvidedProps })}
        </ShallowWapper>
      );

      const googleMapProps = renderPropsWrapper
        .find('[testID="GoogleMaps"]')
        .props();

      expect(googleMapProps.boundingBox).toEqual({
        northEast: {
          lat: 10,
          lng: 12,
        },
        southWest: {
          lat: 12,
          lng: 14,
        },
      });
    });

    it("expect to use currentRefinement when it's defined and hits are not empty", () => {
      const props = {
        ...defaultProps,
      };

      const renderProvidedProps = {
        ...defaultRenderProvidedProps,
        hits: [
          { _geoloc: { lat: 10, lng: 12 } },
          { _geoloc: { lat: 12, lng: 14 } },
        ],
        currentRefinement: {
          northEast: {
            lat: 10,
            lng: 12,
          },
          southWest: {
            lat: 12,
            lng: 14,
          },
        },
      };

      const renderPropsWrapper = shallow(
        <ShallowWapper>
          {renderProps({ props, renderProvidedProps })}
        </ShallowWapper>
      );

      const googleMapProps = renderPropsWrapper
        .find('[testID="GoogleMaps"]')
        .props();

      expect(googleMapProps.boundingBox).toEqual({
        northEast: {
          lat: 10,
          lng: 12,
        },
        southWest: {
          lat: 12,
          lng: 14,
        },
      });
    });

    it("expect to use currentRefinement when it's not defined and hits are empty", () => {
      const props = {
        ...defaultProps,
      };

      const renderProvidedProps = {
        ...defaultRenderProvidedProps,
      };

      const renderPropsWrapper = shallow(
        <ShallowWapper>
          {renderProps({ props, renderProvidedProps })}
        </ShallowWapper>
      );

      const googleMapProps = renderPropsWrapper
        .find('[testID="GoogleMaps"]')
        .props();

      expect(googleMapProps.boundingBox).toBe(undefined);
    });
  });
});
