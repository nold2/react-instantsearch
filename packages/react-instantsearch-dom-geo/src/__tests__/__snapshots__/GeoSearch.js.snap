// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GeoSearch expect to render 1`] = `
<GoogleMaps
  google={
    Object {
      "maps": Object {
        "ControlPosition": Object {
          "LEFT_TOP": "left:top",
        },
        "LatLng": [MockFunction],
        "LatLngBounds": [MockFunction],
        "Map": [MockFunction],
        "Marker": [MockFunction],
        "OverlayView": Object {
          "getPanes": [MockFunction],
          "getProjection": [MockFunction],
          "setMap": [MockFunction],
        },
        "event": Object {
          "addListenerOnce": [MockFunction],
        },
      },
    }
  }
  hasMapMoveSinceLastRefine={false}
  initialPosition={
    Object {
      "lat": 0,
      "lng": 0,
    }
  }
  initialZoom={1}
  isRefineOnMapMove={true}
  mapOptions={Object {}}
  refine={[Function]}
  setMapMoveSinceLastRefine={[Function]}
  testID="GoogleMaps"
>
  <div>
    Hello this is the children
  </div>
</GoogleMaps>
`;
